cajita/src/Cajita_FastFourierTransform.hpp

```diff
@@ -23,291 +23,509 @@
+            (int)entity_space.extent( Dim::I ) };
 
-      0: reshape direct from pencil to pencil
-      1: two reshapes from pencil to brick, then brick to pencil
+        // Get the global grid.
+        const auto& global_grid = layout.localGrid()->globalGrid();
 
-      If this function is not used to set the type then 0 is used as the
-      default.
-    */
-    FastFourierTransformParams& setExchangeType( const int type )
-    {
-        exchange = type;
-        return *this;
+        // Get the low corner of the global index space on this rank.
+        global_low = { (int)global_grid.globalOffset( Dim::K ),
+                       (int)global_grid.globalOffset( Dim::J ),
+                       (int)global_grid.globalOffset( Dim::I ) };
+
+        // Get the high corner of the global index space on this rank.
+        global_high = { global_low[Dim::I] + local_num_entity[Dim::I] - 1,
+                        global_low[Dim::J] + local_num_entity[Dim::J] - 1,
+                        global_low[Dim::K] + local_num_entity[Dim::K] - 1 };
     }
 
     /*!
-      \brief Set the pack type.
-      \param type Pack type.
+      \brief Ensure the FFT compute array matches the FFT class array layout.
+      \param x The array to compare to class data.
+    */
+    template <class Array_t>
+    void checkArray( const Array_t& x )
+    {
+        static_assert( is_array<Array_t>::value, "Must use an array" );
+        static_assert(
+            std::is_same<typename Array_t::entity_type, entity_type>::value,
+            "Array entity type mush match transform entity type" );
+        static_assert(
+            std::is_same<typename Array_t::mesh_type, mesh_type>::value,
+            "Array mesh type mush match transform mesh type" );
+        static_assert(
+            std::is_same<typename Array_t::device_type, device_type>::value,
+            "Array device type and transform device type are different." );
+        static_assert(
+            std::is_same<typename Array_t::value_type,
+                         Kokkos::complex<value_type>>::value ||
+                std::is_same<typename Array_t::value_type, value_type>::value,
+            "Array value type and complex transform value type are "
+            "different." );
 
-      0: array
-      1: pointer
-      2: memcpy
+        if ( 1 != x.layout()->dofsPerEntity() )
+            throw std::logic_error(
+                "Only 1 complex value per entity allowed in FFT" );
+    }
 
-      If this function is not used to set the type then 2 is used as the
-      default.
+    /*!
+      \brief Do a forward FFT.
+      \param x The array on which to perform the forward transform.
+      \param scaling Method of scaling data.
     */
-    FastFourierTransformParams& setPackType( const int type )
+    template <class Array_t, class ScaleType>
+    void forward( const Array_t& x, const ScaleType scaling )
     {
-        packflag = type;
-        return *this;
+        checkArray( x );
```

Comment:
A better way to do this would be to write some kind of meta type checker (i.e a concept) instead of having a function which only does compile-time work. You could simply make all those `is_*` checks in the `static_assert` calls another type checker that would inherit from `std::is_true` only if they are all satisfied.

Reply:
Done